packages/@aws-cdk/aws-eks/lib/helm-chart.ts

```diff
@@ -93,31 +97,49 @@ export class HelmChart extends Construct {
       }
     });
   }
+}
 
-  private getOrCreateHelmChartHandler(cluster: Cluster): lambda.IFunction | undefined {
-    if (!cluster.kubectlEnabled) {
-      return undefined;
-    }
+class HelmResourceProvider extends NestedStack {
+  /**
+   * Creates a stack-singleton resource provider nested stack.
+   */
+  public static getOrCreate(scope: Construct) {
+    const stack = Stack.of(scope);
+    const uid = '@aws-cdk/aws-eks.HelmResourceProvider';
+    return stack.node.tryFindChild(uid) as HelmResourceProvider || new HelmResourceProvider(stack, uid);
+  }
 
-    let handler = cluster.node.tryFindChild('HelmChartHandler') as lambda.IFunction;
-    if (!handler) {
-      handler = new lambda.Function(cluster, 'HelmChartHandler', {
-        code: lambda.Code.fromAsset(path.join(__dirname, 'helm-chart')),
-        runtime: lambda.Runtime.PYTHON_3_7,
-        handler: 'index.handler',
-        timeout: Duration.minutes(15),
-        layers: [ KubectlLayer.getOrCreate(this, { version: "2.0.0-beta1" }) ],
-        memorySize: 256,
-        environment: {
-          CLUSTER_NAME: cluster.clusterName,
-        },
-
-        // NOTE: we must use the default IAM role that's mapped to "system:masters"
-        // as the execution role of this custom resource handler. This is the only
-        // way to be able to interact with the cluster after it's been created.
-        role: cluster._defaultMastersRole,
-      });
-    }
-    return handler;
+  /**
+   * The custom resource provider.
+   */
+  public readonly provider: cr.Provider;
+
+  /**
+   * The IAM role used to execute this provider.
+   */
+  public readonly role: iam.IRole;
+
+  constructor(scope: Construct, id: string) {
```

Comment:
Should this be `private`? I reckon people ought to use `getOrCreate` and only that...

Reply:
Yap, changed.